crates/bevy_render/src/camera/camera.rs

```diff
@@ -343,19 +415,84 @@ impl Camera {
     /// To get the world space coordinates with the viewport position, you should use
     /// [`world_to_viewport`](Self::world_to_viewport).
     ///
-    /// Returns `None` if the `camera_transform`, the `world_position`, or the projection matrix defined by [`CameraProjection`] contain `NAN`.
     /// Panics if the projection matrix is null and `glam_assert` is enabled.
-    pub fn ndc_to_world(&self, camera_transform: &GlobalTransform, ndc: Vec3) -> Option<Vec3> {
+    pub fn ndc_to_world(
+        &self,
+        camera_transform: &GlobalTransform,
+        ndc: Vec3,
+    ) -> Result<Vec3, NdcToWorldError> {
+        let camera_transform_matrix = Self::finite_camera_transform_matrix(camera_transform)
+            .map_err(NdcToWorldError::CameraTransformNotFiniteError)?;
+
+        let projection_matrix = self
+            .finite_projection_matrix()
+            .map_err(NdcToWorldError::ProjectionMatrixNotFiniteError)?;
+
         // Build a transformation matrix to convert from NDC to world space using camera data
-        let ndc_to_world =
-            camera_transform.compute_matrix() * self.computed.projection_matrix.inverse();
+        let ndc_to_world = camera_transform_matrix * projection_matrix.inverse();
 
         let world_space_coords = ndc_to_world.project_point3(ndc);
+        if !world_space_coords.is_finite() {
+            return Err(NdcToWorldError::WorldSpaceCoordsNotFiniteError(
+                world_space_coords,
+            ));
+        }
 
-        (!world_space_coords.is_nan()).then_some(world_space_coords)
+        Ok(world_space_coords)
     }
 }
 
+#[derive(Debug)]
+pub enum NdcToWorldError {
+    WorldSpaceCoordsNotFiniteError(Vec3),
+    CameraTransformNotFiniteError(CameraTransformNotFiniteError),
+    ProjectionMatrixNotFiniteError(BadProjectionMatrixError),
+}
+
+#[derive(Debug)]
+pub enum WorldToNdcError {
+    NdcSpaceCoordsNotFiniteError(Vec3),
+    CameraTransformNotFiniteError(CameraTransformNotFiniteError),
+    ProjectionMatrixNotFiniteError(BadProjectionMatrixError),
+}
+
+#[derive(Debug)]
+pub struct LogicalViewportSizeError {
+    pub viewport_is_set: bool,
+    pub target_info_is_set: bool,
+}
+
+#[derive(Debug)]
+pub enum ViewportToWorldError {
+    LogicalViewportSizeError(LogicalViewportSizeError),
+    CameraTransformNotFiniteError(CameraTransformNotFiniteError),
+    ProjectionMatrixNotFiniteError(BadProjectionMatrixError),
+    WorldNearPlaneNotFiniteError(Vec3),
+    WorldFarPlaneNanError(Vec3),
+}
+
+#[derive(Debug)]
+pub enum ViewportToWorld2DError {
+    NdcToWorldError(NdcToWorldError),
+    LogicalViewportSizeError(LogicalViewportSizeError),
+}
+
+#[derive(Debug)]
+pub enum WorldToViewportError {
+    WorldToNdcError(WorldToNdcError),
+    LogicalViewportSizeError(LogicalViewportSizeError),
+    NdcCoordsOutsideFrustumError(Vec3),
+}
+
+#[derive(Debug)]
+pub enum BadProjectionMatrixError {
+    BadProjectionMatrixError(Mat4),
+    ProjectionMatrixUndefinedError,
```

Comment:
We shouldn't repeat `Error`. Probably only include it on the enum type, not the variants.

Reply:
Fixed